braindecode/models/functions.py

```diff
@@ -45,3 +45,104 @@ def transpose_time_to_spat(x):
         tensor in which last and first dimensions are swapped
     """
     return x.permute(0, 3, 2, 1)
+
+
+def modeig_forward(X, function_applied):
+    """Modified eigenvalue of symmetric matrix X."""
+    s, U = torch.linalg.eigh(X)
+    s_modified = function_applied(s)
+    output = U @ torch.diag_embed(s_modified) @ U.transpose(-1, -2)
+    return output, s, U, s_modified
+
+
+def modeig_backward(grad_output, s, U, s_modified, derivative, threshold=1e-4):
+    """Backward pass of modified eigenvalue of symmetric matrix X."""
+    # compute Loewner matrix
+    denominator = s.unsqueeze(-1) - s.unsqueeze(-1).transpose(-1, -2)
+    is_eq = denominator.abs() < threshold
+    denominator[is_eq] = 1.0
+
+    # case: sigma_i != sigma_j
+    numerator = s_modified.unsqueeze(-1) - s_modified.unsqueeze(-1).transpose(-1, -2)
+
+    # case: sigma_i == sigma_j
+    s_derivativated = derivative(s)
+    numerator[is_eq] = 0.5 * (
+        s_derivativated.unsqueeze(-1) + s_derivativated.unsqueeze(-1).transpose(-1, -2)
+    )[is_eq]
+    L = numerator / denominator
+
+    grad_input = U @  (L * (U.transpose(-1, -2) @ grad_output @ U)) @ U.transpose(-1, -2)
+
+    return grad_input
+
+
+class logm(torch.autograd.Function):
+    """Matrix logarithm of a symmetric matrix.
+
+    This class computes the matrix logarithm of a symmetric matrix X.
+    It also adapt the backpropagation according to the chain rule [1]_.
+
+    Parameters
+    ----------
+    X : torch.Tensor
+        Symmetric matrix
+
+    Returns
+    -------
+    torch.Tensor
+        Matrix logarithm of X
+
+    References
+    ----------
+    .. [1] Brooks et al. 2019, Riemannian batch normalization
+            for SPD neural networks, NeurIPS
+    """
+    @staticmethod
+    def forward(ctx, X):
+        def function_applied(s): return s.log()
+        output, s, U, s_modified = modeig_forward(X, function_applied)
+        ctx.save_for_backward(s, U, s_modified)
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        s, U, s_modified = ctx.saved_tensors
+        def derivative(s): return s.reciprocal()
+        return modeig_backward(grad_output, s, U, s_modified, derivative)
+
+
+class regm(torch.autograd.Function):
+    """Regularized matrix logarithm of a symmetric matrix.
```

Comment:
it's not regularized in the classic sense where you add a constant on the diag. It puts all values less then threshold to 0 so it outputs rank deficient matrices unless I miss something.